cc/algorithms/partition-selection-strategies.h

```diff
@@ -0,0 +1,345 @@
+     absl::optional<int> max_partitions_contributed_;
+
+     //Checks if the max number of partitions contributed  to is set and valid.
+     base::Status MaxPartitionsContributedIsSetAndValid() {
+       if (!max_partitions_contributed_.has_value()) {
+         return base::InvalidArgumentError(
+          "Max number of partitions a user can contribute to has to be set.");
+       }
+       if (max_partitions_contributed_ <= 0) {
+         return base::InvalidArgumentError(absl::StrCat(
+           "Max number of partitions a user can contribute to has to be"
+           " positive but is ",
+           delta_.value()));
+       }
+       return base::OkStatus();
+     }
+
+     // Checks if delta is set and valid.
+     base::Status DeltaIsSetAndValid() {
+       if (!delta_.has_value()) {
+         return base::InvalidArgumentError("Delta has to be set.");
+       }
+       if (!std::isfinite(delta_.value())) {
+         return base::InvalidArgumentError(
+           absl::StrCat("Delta has to be finite but is ", delta_.value()));
+       }
+       if (delta_.value() <= 0 || 1 <= delta_.value()) {
+         return base::InvalidArgumentError(absl::StrCat(
+           "Delta has to be in the interval (0,1) but is ",
+           delta_.value()));
+        }
+        return base::OkStatus();
+      }
+  };
+
+  PartitionSelectionStrategy(double epsilon, double delta,
+                             int max_partitions_contributed)
+  	: epsilon_(epsilon), delta_(delta),
+      adjusted_delta_(1.0 - pow(1 - delta,
+                      (double) max_partitions_contributed)),
+      max_partitions_contributed_(max_partitions_contributed) {}
+
+  virtual ~PartitionSelectionStrategy() = default;
+
+  double GetEpsilon() { return epsilon_; }
+
+  double GetDelta() { return delta_; }
+
+  double GetAdjustedDelta() { return adjusted_delta_; }
+
+  double GetMaxPartitionsContributed() { return max_partitions_contributed_; }
+
+  virtual bool ShouldKeep(int num_users) = 0;
+
+ protected:
+  double epsilon_;
+  double delta_;
+  double adjusted_delta_;
+  int max_partitions_contributed_;
+};
+
+
+class PreaggPartitionSelection : public PartitionSelectionStrategy {
+ public:
+  //Builder for PreaggPartitionSelection
+  class Builder : public PartitionSelectionStrategy::Builder {
+   public:
+     base::StatusOr<std::unique_ptr<PartitionSelectionStrategy>>
+     Build() override {
+     base::Status epsilon_status = EpsilonIsSetAndValid();
+     base::Status delta_status = DeltaIsSetAndValid();
+     base::Status max_partitions_contributed_status =
+       MaxPartitionsContributedIsSetAndValid();
+
+     if (!epsilon_status.ok()) {
+       return epsilon_status;
+     }
+     else if (!delta_status.ok()) {
+       return delta_status;
+     }
+     else if (!max_partitions_contributed_status.ok()) {
+       return max_partitions_contributed_status;
+     }
+     else {
+       std::unique_ptr<PartitionSelectionStrategy> magic = 
```

Comment:
nit: You can directly return and avoid assigning to `magic` (which could otherwise have a better name).

Reply:
Actually, returning directly will cause an error.  I think it has to do with the unique pointer going out of scope?  Also the name magic is from the name of this method of partition selection, "magic partition selection."  I guess magic does sound kind of vague though, so I renamed it to magic_selection.  Hopefully that fixes it!

Reply:
There seems to be something missing about our forked version of StatusOr that makes us have to assign to a variable before returning. We have to do this is LaplaceMechanism too.